src/lib.rs

```diff
@@ -13,24 +13,103 @@
 // under the License.
 
 extern crate fuse;
+extern crate libc;
 #[macro_use] extern crate log;
+extern crate time;
 
+use std::collections::HashMap;
 use std::ffi::OsStr;
 use std::io;
 use std::path::Path;
+use std::sync::{Arc, Mutex};
+use time::Timespec;
+
+mod nodes;
+
+// TODO(jmmv): Make configurable via a flag and store inside SandboxFS.
+pub const TTL: Timespec = Timespec { sec: 60, nsec: 0 };
 
 /// FUSE file system implementation of sandboxfs.
 struct SandboxFS {
+    /// Mapping of inode numbers to in-memory nodes that tracks all files known by sandboxfs.
+    nodes: Arc<Mutex<HashMap<u64, Arc<nodes::Node>>>>,
 }
 
 impl SandboxFS {
     /// Creates a new `SandboxFS` instance.
     fn new() -> SandboxFS {
-        SandboxFS {}
+        let root = {
+            let inode = fuse::FUSE_ROOT_ID;
+            let now = time::get_time();
+            let uid = unsafe { libc::getuid() } as u32;
+            let gid = unsafe { libc::getgid() } as u32;
+            nodes::Dir::new_empty(inode, inode, now, uid, gid)
+        };
+
+        let mut nodes = HashMap::new();
+        nodes.insert(root.inode(), root);
+        SandboxFS {
+            nodes: Arc::from(Mutex::from(nodes)),
+        }
+    }
+
+    /// Gets a node given its `inode`.
+    ///
+    /// We assume that the inode number is valid and that we have a known node for it; otherwise,
+    /// we crash.  The rationale for this is that this function is always called on inode numbers
+    /// requested by the kernel, and we can trust that the kernel will only ever ask us for inode
+    /// numbers we have previously told it about.
+    fn find_node(&mut self, inode: u64) -> Arc<nodes::Node> {
+        let nodes = self.nodes.lock().unwrap();
+        match nodes.get(&inode) {
+            Some(node) => node.clone(),
+            None => panic!("Kernel requested unknown inode {}", inode),
+        }
     }
 }
 
 impl fuse::Filesystem for SandboxFS {
+    fn getattr(&mut self, _req: &fuse::Request, inode: u64, reply: fuse::ReplyAttr) {
+        let node = self.find_node(inode);
+        match node.getattr() {
+            Ok(attr) => reply.attr(&TTL, &attr),
+            Err(e) => reply.error(e.raw_os_error().unwrap()),
+        }
+    }
+
+    fn lookup(&mut self, _req: &fuse::Request, parent: u64, name: &OsStr, reply: fuse::ReplyEntry) {
+        let dir_node = self.find_node(parent);
+        match dir_node.lookup(name) {
+            Ok((node, attr)) => {
+                {
+                    let mut nodes = self.nodes.lock().unwrap();
+                    if !nodes.contains_key(&node.inode()) {
```

Comment:
This is more commonly written `nodes.entry(nodes.inode()).or_insert(node);`.

Reply:
Done.